server/src/main/java/org/elasticsearch/action/search/FetchLookupFieldsPhase.java

```diff
@@ -0,0 +1,132 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Asynchronously resolves {@link LookupField} that are specified {@link DocumentField#getLookupFields()}
+ *
+ * @see org.elasticsearch.index.mapper.LookupRuntimeFieldType
+ */
+final class FetchLookupFieldsPhase extends SearchPhase {
+    private final SearchPhaseContext context;
+    private final InternalSearchResponse searchResponse;
+    private final AtomicArray<SearchPhaseResult> queryResults;
+
+    FetchLookupFieldsPhase(SearchPhaseContext context, InternalSearchResponse searchResponse, AtomicArray<SearchPhaseResult> queryResults) {
+        super("fetch_lookup_fields");
+        this.context = context;
+        this.searchResponse = searchResponse;
+        this.queryResults = queryResults;
+    }
+
+    private record Cluster(String clusterAlias, List<SearchHit> hitsWithLookupFields, List<LookupField> lookupFields) {
+
+    }
+
+    private static List<Cluster> groupLookupFieldsByClusterAlias(InternalSearchResponse response) {
+        final Map<String, List<SearchHit>> perClusters = new HashMap<>();
+        for (SearchHit hit : response.hits.getHits()) {
+            String clusterAlias = hit.getClusterAlias() != null ? hit.getClusterAlias() : RemoteClusterService.LOCAL_CLUSTER_GROUP_KEY;
+            if (hit.hasLookupFields()) {
+                perClusters.computeIfAbsent(clusterAlias, k -> new ArrayList<>()).add(hit);
+            }
+        }
+        final List<Cluster> clusters = new ArrayList<>(perClusters.size());
+        for (Map.Entry<String, List<SearchHit>> e : perClusters.entrySet()) {
+            final List<LookupField> lookupFields = e.getValue()
+                .stream()
+                .flatMap(h -> h.getDocumentFields().values().stream())
+                .flatMap(doc -> doc.getLookupFields().stream())
+                .distinct()
+                .toList();
+            clusters.add(new Cluster(e.getKey(), e.getValue(), lookupFields));
+        }
+        return clusters;
+    }
+
+    @Override
+    public void run() {
+        final List<Cluster> clusters = groupLookupFieldsByClusterAlias(searchResponse);
+        if (clusters.isEmpty()) {
+            context.sendSearchResponse(searchResponse, queryResults);
+            return;
+        }
+        final MultiSearchRequest multiSearchRequest = new MultiSearchRequest();
+        for (Cluster cluster : clusters) {
+            // Do not prepend the clusterAlias to the lookupIndex if the search request is already on the remote cluster.
+            final String clusterAlias = context.getRequest().getLocalClusterAlias() == null ? cluster.clusterAlias : null;
+            for (LookupField lookupField : cluster.lookupFields) {
+                final SearchRequest searchRequest = lookupField.toSearchRequest(clusterAlias);
+                searchRequest.setCcsMinimizeRoundtrips(context.getRequest().isCcsMinimizeRoundtrips());
```

Comment:
Maybe it'd be clearer to assert that ccs_minimize_roundtrips is always 'false' if we're hitting any remote cluster. Because if it's 'true', then we would have already handled the lookup in the remote call.

Reply:
+1, I added the assertion in https://github.com/elastic/elasticsearch/pull/82385/commits/be08cca75326888f596e41e2b882695a78da3257

Reply:
Great, but then here we could do just use `false` here to be clear? Because `TransportSearchAction.shouldMinimizeRoundtrips` can return a different response than `context.getRequest().isCcsMinimizeRoundtrips()`.

Reply:
I think using the ccs_minimize_roundtrips of the original search request is closer to the user's intention. But I am ok with using false. I've adjusted this in https://github.com/elastic/elasticsearch/pull/82385/commits/5f8a3a2c5531a3b0d068ca559f71430896c4d075.